### ISSUE TITLE
Ease memory pressure by deprioritizing root tasks?

### ISSUE THREAD

**Issue by TomNicholas** [dask/distributed#6360]
The need for [memory backpressure](https://github.com/dask/distributed/issues/2602) has been discussed at length, and a full solution is arguably necessary to solve many of the use cases in the pangeo community. Whilst [efforts in that direction](https://github.com/dask/distributed/pull/5251#issuecomment-1124389237) do seem be underway, I would like to suggest a simpler stop-gap solution for the shorter term. The basic issue is that workers can overeargerly open new data at a rate faster than other tasks free up data, and is best understood via the figure under "root task overproduction" [here](https://coiled.io/blog/better-shuffling-in-dask-a-proof-of-concept/).

All my tasks that consume memory are root tasks, so **is there any way to deprioritize root tasks**? This should be a lot simpler than solving the general problem, because it doesn't require (a) any records kept of how much memory each task uses or (b) workers to change their priorities over the course of a computation based on new information. All I need is for workers to be discouraged / prevented from opening new data until the task chains that are already begun are as complete as possible. This will obviously waste time as workers wait, but I would be happy to accept a considerable slowdown if it means that I definitely won't run out of memory.

This problem applies to any computation where the data is largest when opened and then steadily reduced (which is most pangeo workflows, with the important exception of the groupby stuff IIUC), but the opening task and the memory-relieving tasks can't be directly fused into one.

Is there some way of implementing this? I'm happy to provide an example use case if that would help, or to try hacking away at the distributed code, but I wanted to know if this is even possible first.

cc @gjoseph92 , and @rabernat @dcherian @TomAugspurger , because this is related to the discussion we had in the pangeo community meeting the other week.

**Comment by gjoseph92** [comment]
The figure @TomNicholas was referencing (thanks for the reminder, I forgot about this!):
![image](https://user-images.githubusercontent.com/3309802/168930605-9a609bd8-6ba7-4ef8-9445-3ec0cb016b6d.png)

I mention that because this problem is inherent to the way the current task scheduling algorithm works in distributed; see https://github.com/dask/distributed/issues/5223 and https://github.com/dask/distributed/issues/5555.. So approaches like https://github.com/dask/distributed/issues/4891 are not, in my mind, the "full solution"—they'd be layering on more complexity to an already-complex algorithm with two components (normal eager scheduling, and task stealing aka occasional rebalancing). "Speculative task assignment" (https://github.com/dask/distributed/issues/3974) is more what I'd consider the "full solution", thought I think there are other approaches that could work as well.

Just for context :)

**Comment by gjoseph92** [comment]
> so is there any way to deprioritize root tasks?

As a way to hack around this, right now, I'd play with [worker resources](https://distributed.dask.org/en/stable/resources.html). For example, something along the lines of (untested):

```python
import dask
import coiled
import xarray as xr

NTHREADS = 4
cluster = coiled.Cluster(..., worker_cpu=NTHREADS, worker_options={"resources": {"ROOT": NTHREADS}})
client = dask.distributed.client(cluster)

with dask.annotate(resources={"ROOT": 1}):
    ds = xr.open_dataset(...)

result = ds.a - ds.a.mean()
with dask.config.set({"optimization.fuse.active": False}):
    result.compute()
```

A few caveats:
* I feel like I've tried this before and it didn't work that well.
* Tasks with resource restrictions won't benefit from [root task colocation](https://distributed.dask.org/en/stable/scheduling-policies.html#initial-task-placement) https://github.com/dask/distributed/pull/4967, which places root tasks whose outputs will be used together onto the same workers initially, which can massively reduce data transfer in the cluster. Instead, root tasks will be placed essentially randomly. It's possible that you'll avoid root task overproduction, but then have excessive memory use from data transfers.
* The resource annotation will be lost without `"optimization.fuse.active": False`, see https://github.com/dask/dask/issues/7036. These low-level optimization can actually be quite important for array workloads, so losing them may hurt performance.
* Scheduling with resources is known to be buggier:
  * https://github.com/dask/distributed/issues/6137
  * https://github.com/dask/distributed/issues/5937

**Comment by dcherian** [comment]
One way to "solve" this is to batch the computation (e.g. compute the result 100 timesteps at a time) and write to zarr (some ugly code below).
<details>

``` python
def batch_load(obj, factor=2):
    """
    Load xarray object values by calling compute on block subsets (that are an integral multiple of chunks along each chunked dimension)

    Parameters
    ----------
    obj: xarray object
    factor: int
        multiple of chunksize to load at a single time.
        Passed on to split_blocks
    """
    if isinstance(obj, xr.DataArray):
        dataset = obj._to_temp_dataset()
    else:
        dataset = obj


    # result = xr.full_like(obj, np.nan).load()
    computed = []
    for label, chunk in split_blocks(dataset, factor=factor):
        print(f"computing {label}")
        computed.append(chunk.compute())
    result = xr.combine_by_coords(computed)


    if isinstance(obj, xr.DataArray):
        result = obj._from_temp_dataset(result)


    return result




def batch_to_zarr(ds, file, dim, batch_size, restart=False, **kwargs):
    """
    Batched writing of dask arrays to zarr files.

    Parameters
    ----------
    ds : xarray.Dataset
        Dataset to write.
    file : str
        filename
    dim : str
        Dimension along which to split dataset and append. Passed to `to_zarr` as `append_dim`
    batch_size : int
        Size of a single batch

    Returns
    -------
    None
    """


    import tqdm


    if not restart:
        ds.isel({dim: [0]}).to_zarr(file, consolidated=True, mode="w", **kwargs)
    else:
        print("Restarting...")
        opened = xr.open_zarr(file, consolidated=True)
        ds = ds.sel(time=slice(opened[dim][-1], None))
        print(
            f"Last index = {opened[dim][-1].values}. Starting from {ds[dim][1].values}"
        )
        opened.close()


    for t in tqdm.tqdm(range(1, ds.sizes[dim], batch_size)):
        if "encoding" in kwargs:
            del kwargs["encoding"]
        ds.isel({dim: slice(t, t + batch_size)}).to_zarr(
            file, consolidated=True, mode="a", append_dim=dim, **kwargs
        )
```
</details>

Effectively this prevents the scheduler from ever seeing "too many" root tasks.

This now reminds me of the rechunker approach
![image](https://user-images.githubusercontent.com/2448579/168940329-9eb1e749-a7e7-41e4-83bf-6ca9677b8c84.png)

where the dummy task in the middle prevents scheduler from getting too far ahead, and consequently limits memory usage.

So perhaps there can be a new function in `graph_manipulation.py` that takes as input an integer `N` and a task token `token`. This should make the `N+1:2N`th tasks execute only after the `N`th task has completed executing.

**Comment by TomNicholas** [comment]
Thanks both for your replies.

@dcherian batching like you suggested is one of my backup plans :sweat_smile:, but thank you for the code example!

@gjoseph92 I tried playing with worker resources as you suggested, but if it made any difference to the behaviour I couldn't tell. 

Without setting a limit on resources my cluster starts failing like this:

![Screenshot from 2022-05-26 14-37-46](https://user-images.githubusercontent.com/35968931/170581624-7178c373-9284-44b6-9f4a-fcfbda0df6c2.png)

and after I set a limit via

```python
options = g.cluster_options()
NTHREADS = 1
options.worker_cores = NTHREADS
options.worker_memory = 50

with dask.config.set({"resources": {"ROOT": NTHREADS}}):
    gc = g.new_cluster(cluster_options=options)

with dask.annotate(resources={"ROOT": 1}):
    ds = xr.open_dataset(...)

omega = vort(ds)  # my calculation, an xarray.apply_ufunc

with dask.config.set({"optimization.fuse.active": False}):
    with gc.get_client():
        # write result to zarr
        st = omega.to_dataset(name="vort").to_zarr(mapper, mode='w', compute=True)
```
it looks like this

![Screenshot from 2022-05-26 16-33-31](https://user-images.githubusercontent.com/35968931/170581647-7c1ba964-d0f2-46a8-98d0-1789d025a0fd.png)

You can see that in both cases it has managed to save out ~300 chunks (`store_map`), but the data spilled to disk still keeps steadily growing.

---

At @jbusecke's suggestion I also tried setting the task priority via

```python
with dask.annotate(priority=-999999):
    ds = xr.open_dataset(...)

with dask.annotate(priority=999999):
    with gc.get_client():
        # write result to zarr
        st = omega.to_dataset(name="vort").to_zarr(mapper, mode='w', compute=True)
```

which appeared to help the computation proceed a bit somewhat, but after a while my memory usage still keeps growing indefinitely.

---

This is a subset of my optimised (partially fused) task graph, for two timesteps. It's basically two `open_dataset` calls per timestep, each of which is padded with xarray, then fed into a simple `apply_ufunc` which returns one chunk, then written to a single zarr chunk.

![image](https://user-images.githubusercontent.com/35968931/170579587-643d5c56-5661-459e-9f73-8108a40202b9.png)

At full scale I'm trying to compute 9000 timesteps. The difference is that the outer-most 4 task chains (2 on each side) are duplicated (there are 180000 `open`'s for the full problem), but the middle 3 chains don't change (those are grid-related variables that don't depend on time), so it's basically 2N+3 `open_dataset` calls for N time chunks.

The static ordering looks reasonable - dark orange should be written out before yellow and light orange are loaded

![image](https://user-images.githubusercontent.com/35968931/170580234-9b313e99-2b30-4427-859f-c698e1252bdc.png)

But this static ordering seems to be largely irrelevant to what actually happens when you have multiple workers.

---

Is this a pathological fail case? Whatever I've tried my 20 workers still race to open ~400 chunks (loading hundreds of GBs into memory and spilling to disk) before they manage write a single chunk out, even though each output chunk only requires two input chunks.

There might be some way for me to inline/fuse my way to an embarrassingly parallel graph, so that the open and write steps are in one task, but at that point I may as well have just used `map_blocks` (or a different parallel execution framework).

**Comment by gjoseph92** [comment]
> But this static ordering seems to be largely irrelevant to what actually happens when you have multiple workers

Yes, the priority ordering doesn't have that much effect (it's just a tie-breaker, not an actual ordering that the scheduler follows).
* https://github.com/dask/distributed/issues/5555

> if it made any difference to the behaviour I couldn't tell

I'm a bit surprised by this (but I agree, it didn't look like anything changed).

I looked into it a little and it turns out the scheduling of tasks that use resources doesn't work the way I expected:
* https://github.com/dask/distributed/issues/6467

https://github.com/dask/distributed/pull/6468 makes it work the way you'd expect. @TomNicholas, I'd be really curious to see how things work if you keep using resources, but switch to that PR. There's been lots of discussion of root-task overproduction, but we've never actually gotten to see how things would go if it didn't happen, so having this comparison would be very valuable.

**Comment by TomNicholas** [comment]
Thanks so much for that @gjoseph92 - I'm excited to try it out, but it'll take me a few days as I need to work out how to run on a custom version of `distributed` first.

**Comment by gjoseph92** [comment]
You could use Coiled if that would be easier; you can install `git+https://github.com/gjoseph92/distributed.git@hold-back-tasks-without-available-resources` in your [software environment](https://docs.coiled.io/user_guide/software_environment_creation.html#pip).

**Comment by gjoseph92** [comment]
@TomNicholas why don't you join the Coiled slack; it'll be easier to go back and forth there. I think this link works, but if not it's at the bottom-left of https://cloud.coiled.io: https://join.slack.com/t/coiled-users/shared_invite/zt-hx1fnr7k-In~Q8ui3XkQfvQon0yN5WQ.

I also haven't tested my PR at all. It's entirely possible I messed up something silly. The cluster logs might show more. I also wouldn't be surprised if you need to add `git+https://github.com/dask/dask.git@main`, though I thought distributed might pull this in automatically.

**Comment by jbusecke** [comment]
> I also wouldn't be surprised if you need to add git+https://github.com/dask/dask.git@main, though I thought distributed might pull this in automatically.

We actually tried this actually, but it failed due a dependency pinned on the last dask release. Just FYI

**Comment by gjoseph92** [comment]
@TomNicholas and I reproduced a variant of his failing workload on Coiled, running on my PR https://github.com/dask/distributed/issues/6467. We used worker resources to annotate root tasks as described in https://github.com/dask/distributed/issues/6360#issuecomment-1129445795. With my PR, this made it so the scheduler would stream root tasks into workers, only submitting a new root task when it heard the previous had completed, instead of submitting all root tasks up front, which the current scheduling algorithm does.

This allowed us to simulate how a cluster would perform if root task overproduction were fixed. The results are compelling:
![plot](https://user-images.githubusercontent.com/3309802/172930466-c62f68a2-d303-4582-98be-8e8a704289f9.png)

**Withholding root tasks from workers enormously reduced memory use (15x lower peak memory usage) while cutting runtime nearly in half.**

_These trials used the same workload on the same cluster, both on https://github.com/dask/distributed/issues/6467. The only difference was "root tasks withheld" wrapped `xr.open_dataset` in `dask.annotate(resources={"ROOT": 1})`, and "standard" did not. All workers were `r5.2xlarge`, with 1 `ROOT` resource each._

You can see how different memory usage looked on the dashboard:

![dashboards](https://user-images.githubusercontent.com/3309802/172944896-68f181df-7299-4e91-b7bc-e917a14aa9a0.png)

It's important to note that because we used resource restrictions, scheduling didn't use co-assignment logic (https://github.com/dask/distributed/pull/4967). I believe that's the reason you see transfers on the second dashboard (two tasks that contributed to the same one output were scheduled on different workers). Implemented properly to take advantage of this, I expect fixing root task overproduction would give even better runtime (maybe 20-40% faster?) and lower memory usage than what we see here.

----

Important takeaways:
* Root task overproduction is a critical issue, hampering memory usage and therefore both runtime and stability.
* It could be addressed just on the scheduler side, without the complexity of https://github.com/dask/distributed/issues/3974, and still have far better performance than current scheduling.

I estimate that a scheduler-only fix for root task overproduction, which just withheld root tasks from workers while maintaining our co-assignment logic, could be implemented in a week.

cc @fjetter @hayesgb @mrocklin

**Comment by gjoseph92** [comment]
Part of the week includes coming up the the design. But the overall idea is something like:

* Identify root-ish tasks using something like the [logic we already use](https://github.com/dask/distributed/blob/43ca938161ce6c3aa6c2886cd5da2438310ec1e4/distributed/scheduler.py#L1768-L1775)
* Don't have more than `SchedulerState.total_nthreads` of these root-ish tasks in `processing` at once. Hold back the extras on the scheduler somewhere, in either state `no-worker` or a new state like `withheld` (or `ready`, with the same meaning as on the worker).

The implementation of this is what needs consideration, particularly how we still get co-assignment (which right now relies on running through all the root tasks at once). Maybe something like:
* `transition_waiting_processing` recommends the task to `withheld` instead of `processing` if `ts.rootish and len(selected_worker.processing) >= selected_worker.nthreads`.
* It stores the planned worker assignment on `ts.planned_worker` (this is hacky and I don't like it, but it's an easy way to maintain co-assignment)
* It tracks these withheld tasks in a data structure like priority heap
* When any rootish task completes, we pop the next task off the heap and recommend scheduling it as part of `_remove_from_processing`

**Comment by mrocklin** [comment]
Historically I would have said "I'd rather not complicate the scheduler further.  It makes sense to have the worker handle this given how relatively simple the worker is"  With recent changes though I'm not sure that that is still the case.  It seems like the worker might be a place of more complexity than the scheduler these days.  I neither love nor hate this plan.

**Comment by gjoseph92** [comment]
The scheduler-only aspect of it is the main appeal to me, for that reason. The worker also can't do this on its own without extra information from the scheduler (it doesn't know enough about overall graph structure). So we either have to touch both scheduler and worker, or only scheduler.

**Comment by mrocklin** [comment]
It could guess.  We could pass rootishness around.  It could also see that "hey, I often have tasks that arrive just after these rootish tasks are done.  I guess I should maybe wait a bit".  I'm not pushing for this though.  The worker seems to be a larger source of complexity today than the scheduler.

**Comment by gjoseph92** [comment]
I really don't like the idea of guessing. Passing rootishness around was one of my thoughts, but it introduces consistency issues (if downstream tasks are cancelled, you have to inform workers to de-rootify their tasks) without providing any complexity or performance benefit (without STA, the worker would still have to wait for a scheduler roundtrip before it could run a new task).

**Comment by mrocklin** [comment]
I'm not pushing for my thought above.  There's no need to argue against it.

I do think that you should develop some comfort with guessing though.  We don't need to be perfect, just good in aggregate.

**Comment by fjetter** [comment]
I agree with @gjoseph92 here and I wouldn't want to implement any logic on the worker that would introduce artificial waiting times. It's current state is not well suited to handle this kind of logic well. I do not see any problems with [STA](https://github.com/dask/distributed/issues/3974) on worker side. We likely need to adjust a few transition rules but that should be doable after the latest changes.

---

> Don't have more than SchedulerState.total_nthreads of these root-ish tasks in processing at once

The request of "I would like to limit the total number of tasks of type X concurrently executing on the cluster" has come up a couple of times, see

- https://github.com/dask/distributed/issues/6382
- https://github.com/dask/distributed/issues/4104

I'm wondering if this can/should be solved first generically before we write a specialized version for root tasks. That would obviously only be part of the solution but it would be a re-usable part

> The implementation of this is what needs consideration, particularly how we still get co-assignment

Can we first deliver something that does not care about co-assignment? IIUC, your prototype does not consider co-assignment and it still is a major win.

**Comment by mrocklin** [comment]
> I agree with @gjoseph92 here and I wouldn't want to implement any logic on the worker 

Just to be perfectly clear, so do I.  I brought up the other approach to say "there are other ways we could think about doing this" not to say "we should do it this way instead".

**Comment by gjoseph92** [comment]
Update: we're planning on working on this, starting late next week or the week after: https://github.com/dask/distributed/issues/6560. The goal is a draft PR (not necessarily merged) within 2 weeks. When that's up, we'd very much appreciate input from people involved in this thread who could test us out and let us know how it affects performance on real-world cases.

**Comment by TomNicholas** [comment]
[Here's a notebook](https://gist.github.com/TomNicholas/b4f6a2e839ecea403ab291c9bc0c2c02) containing a simple example demonstrating the root task overproduction problem.

This example **only uses dask**, not importing xarray or xGCM anywhere. The offending graph looks like this


![image](https://user-images.githubusercontent.com/35968931/173446824-80b4fd71-a15e-46dd-84ad-4d0fdd4b4497.png)


I think @gjoseph92 was able to come up with some even simpler examples, but I'm sharing this one because it's the simplest one I could create that still looks somewhat like the original problem I was trying to compute (with xGCM).

**Comment by gjoseph92** [comment]
Two even simpler examples (pure dask). These are `a + b` with just enough complexity to make them not fuse into a single task:
```python
import dask.array as da

a = da.random.random(100, chunks=10)
b = da.random.random(100, chunks=10)
s = da.stack([a, b])
r = s.sum(axis=0)
r.visualize(optimize_graph=True)
```
![mydask](https://user-images.githubusercontent.com/3309802/173449479-f4eb25c4-8909-4eae-ac47-32cac58ce8f1.png)

```python
import dask.array as da

a = da.random.random(100, chunks=10)
b = da.random.random(100, chunks=10)
r = a[1:] + b[1:]
r.visualize(optimize_graph=True)
```
![mydask](https://user-images.githubusercontent.com/3309802/173449537-710d85c3-3272-4a17-a87a-25b5a581f3d1.png)

`sum` and slicing aren't blockwise operations. So graph optimization won't fuse `a`, `b`, and the operation into a single task per chunk—there will be three tasks per chunk.

The input tasks (random array creation) will be overproduced relative to output tasks (`sum`/`add`), leading to excessive memory use.

**Comment by gjoseph92** [comment]
Also, the problem @TomNicholas has been demonstrating here is a simplified form of what he actually needs to solve.

Here, he's computing `a + b` (which runs out of memory due to root task overproduction).

What he actually needs to compute is `broadcast1 * a + broadcast2 * b`. Where `a` and `b` are n-dimensional, and the `broadcast` are (N-1)-dimensional, and broadcast against many chunks of `a` and `b`.

This not only suffers from root task overproduction, but also hits the "widely-shared dependencies dogpile" I tried to address in https://github.com/dask/distributed/pull/5325.
